### PR TITLE
Add docker compose based test environment for Linux

### PR DESCRIPTION
**Pull request by pdgendt**
Updated #687 as it appeared to be stale, but still useful.

### CODE REVIEW

**Comment by pdgendt** [comment]
@marc-hb any objections to this PR? I found it very useful to run these extensive tests locally without the need for a lot of manuals setups.
It doesn't change the source code or packages.

File: docker-testing/compose.yaml [hunk 1]
```diff
@@ -0,0 +1,88 @@
+# It would be nicer to consolidate the common user/volumes
+# boilerplate, but there wasn't enough time to figure out if it was
+# possible at time of writing. Improvements welcome.
```
Comment (marc-hb) [inline docker-testing/compose.yaml:3]: According to the interwebs, YAML anchors should Just Work

https://nickjanetakis.com/blog/docker-tip-82-using-yaml-anchors-and-x-properties-in-docker-compose

YAML anchors are a plain, standard, built-in YAML feature - nothing specific to Docker at all.

Comment (marc-hb) [reply docker-testing/compose.yaml:3]: According to the same interwebs, you can expand anchors with either `yq explose` or `yaml merge-expand` and make sure you got your YAML anchors right. Either will be orders of magnitude faster than re-creating the image!

**Review by marc-hb — COMMENTED** [review]
> It doesn't change the source code or packages.

Short of using a separate repository (admittedly a bit overkill but see why in https://github.com/zephyrproject-rtos/west/pull/687#issuecomment-1751629245), please make sure Docker stuff stays contained in a separate sub directory.

File: docker-testing/compose.yaml (hunk 1 above)
Comment (marc-hb) [reply docker-testing/compose.yaml:3]: Also, `yamllint -f parsable file.yaml` has saved me a LOT of YAML time in the past.

(for automation try something like `yamllint -d "{extends: relaxed, rules: {line-length: {max: 100}}}" file.yaml`)

**Comment by pdgendt** [comment]
@marc-hb reduced to mostly reusable files and yaml sections

File: docker-testing/in-container-test.sh [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    TOX_LOG_IN_HOST="$WEST_TOX_OUT_IN_HOST/tox.log"
+    WEST_TESTDIR="/tmp/west"
+
+    mkdir "$WEST_TOX_OUT"
+
+    git clone -q "$WEST" "$WEST_TESTDIR" || die "failed to clone west to $WEST_TESTDIR in container"
+    cd "$WEST_TESTDIR"
+
+    echo "running tox, output in $TOX_LOG_IN_HOST in host"
+    tox run >"$TOX_LOG" 2>&1 || die "tox failed"
```
Comment (marc-hb) [inline docker-testing/in-container-test.sh:39]:
```suggestion
    tox run >"$TOX_LOG" 2>&1 || die "tox failed, see $TOX_LOG"
```

**Review by marc-hb — APPROVED** [review]

File: docker-testing/in-container-test.sh (hunk 2 above)
Comment (pdgendt) [reply docker-testing/in-container-test.sh]: Done

**Pull request merged**